packages/angular_devkit/build_angular/src/angular-cli-files/plugins/karma.ts

```diff
@@ -155,14 +155,26 @@ const init: any = (config: any, emitter: any, customFileHandlers: any) => {
     throw e;
   }
 
-  ['invalid', 'watch-run', 'run'].forEach(function (name) {
-    compiler.plugin(name, function (_: any, callback: () => void) {
-      isBlocked = true;
+  function handler(callback?: () => void) {
+    isBlocked = true;
 
-      if (typeof callback === 'function') {
-        callback();
-      }
-    });
+    if (typeof callback === 'function') {
+      callback();
+    }
+  }
+
+  const hooks = [
+    {method: 'sync', name: 'invalid'},
+    {method: 'async', name: 'watchRun'},
+    {method: 'async', name: 'run'}
+  ];
+
+  hooks.forEach(function (hook) {
+    if (hook.method === 'sync') {
+      compiler.hooks[hook.name].tap('karma', () => handler());
+    } else {
+      compiler.hooks[hook.name].tapAsync('karma', (_: any, callback: () => void) => handler(callback));
+    }
```

Comment:
While this honors the original code, it appears more complex then necessary considering there are only three.  Can you change this to tap the hooks directly?

Reply:
done